### PR TITLE
710: Workload report

### PR DESCRIPTION
**Pull request by jeppekroghitk**
#### Link to ticket

https://leantime.itkdev.dk/#/tickets/showTicket/710

#### Description

Generate a report that shows percentage of hours logged in relation to predefined workload (normtid) for each employee.
- Add option to view either weeks or months in current year

#### Screenshot of the result

Screenshot 2024-06-07 at 15 19 44

Screenshot 2024-06-07 at 15 51 28

#### Checklist

- [ ] My code is covered by test cases.
- [ ] My code passes our test (all our tests).
- [ ] My code passes our static analysis suite.
- [ ] My code passes our continuous integration process.

### CODE REVIEW

File: migrations/Version20240606085631.php [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    public function getDescription(): string
+    {
+        return '';
+    }
+
+    public function up(Schema $schema): void
+    {
+        // this up() migration is auto-generated, please modify it to your needs
+        $this->addSql('CREATE TABLE worker (id INT AUTO_INCREMENT NOT NULL, email VARCHAR(180) NOT NULL, workload DOUBLE PRECISION NOT NULL, UNIQUE INDEX UNIQ_9FB2BF62E7927C74 (email), UNIQUE INDEX UNIQ_9FB2BF621203AA7B (workload), PRIMARY KEY(id)) DEFAULT CHARACTER SET utf8mb4 COLLATE `utf8mb4_unicode_ci` ENGINE = InnoDB');
+        $this->addSql('ALTER TABLE issue ADD tags VARCHAR(255) DEFAULT NULL, DROP hours_remaining');
```
[the remark this answers is not in the document]
Comment (tuj) [reply migrations/Version20240606085631.php:24]: You need to roll back the database to the state before this feature, remove these two migrations files and run make:migrations again

File: src/Repository/WorklogRepository.php [hunk 2]
```diff
@@ -165,4 +165,19 @@ public function getDistinctWorklogUsers(): array
 
         return $workers;
     }
+
+    public function findWorklogsByWorkerAndDateRange($worker, $date_from, $date_to)
```
Comment (tuj) [inline src/Repository/WorklogRepository.php:169]: Add parameter types

File: src/Service/WorkloadReportService.php [hunk 3]
```diff
@@ -0,0 +1,176 @@
+<?php
+
```
Comment (tuj) [inline src/Service/WorkloadReportService.php:2]: This file looks broken...

**Review by tuj — CHANGES_REQUESTED** [review]

File: migrations/Version20240606085631.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (jeppekroghitk) [reply migrations/Version20240606085631.php]: You are correct. This was the way to do it. It is fixed now

File: src/Service/WorkloadReportService.php [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+    /**
+     * Get the workload report based on the specified view mode.
+     *
+     * @param string $viewMode The view mode (default: 'week')
+     *
+     * @return WorkloadReportData The workload report data
+     *
+     * @throws \Exception When an unexpected value for viewMode is provided
+     */
+    public function getWorkloadReport(string $viewMode = 'week'): WorkloadReportData
```
Comment (tuj) [inline src/Service/WorkloadReportService.php:28]: viewMode should be changed to an enum: https://www.php.net/manual/en/language.types.enumerations.php

File: src/Service/WorkloadReportService.php [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+    {
+        $workloadReportData = new WorkloadReportData($viewMode);
+        $workers = $this->workerRepository->findAll();
+
+        foreach ($periods as $period) {
+            // Get period specific readable period representation for table headers.
+            $readablePeriod = $getReadablePeriod($period);
+            $workloadReportData->period->set($period, $readablePeriod);
+        }
+        foreach ($workers as $worker) {
```
Comment (tuj) [inline src/Service/WorkloadReportService.php:84]: empty lines between for loops

File: src/Service/WorkloadReportService.php [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+
+                if (!$workerWorkload) {
+                    $workerId = $worker->getUserIdentifier();
+                    throw new \Exception("Workload of worker: $workerId cannot be unset when generating workload report.");
+                }
+
+                // Get total logged percentage based on weekly workload.
+                $roundedLoggedPercentage = $this->getRoundedLoggedPercentage($loggedHours, $workerWorkload, $viewMode);
+
+                if (!$roundedLoggedPercentage) {
```
Comment (tuj) [inline src/Service/WorkloadReportService.php:116]: the function getRoundedLoggedPercentage cannot return null according to its return type.

File: src/Service/DateTimeHelper.php [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+
+        return $weekArray;
+    }
+
+    /**
+     * Retrieves an array of months and their corresponding numeric representation of a year.
+     *
+     * @return array an array of months with corresponding numeric representation (1-12)
+     */
+    public function getMonthsOfYear(): array
```
Comment (tuj) [inline src/Service/DateTimeHelper.php:101]: Remove this function and use range instead: https://www.php.net/range

**Review by tuj — CHANGES_REQUESTED** [review]

File: src/Service/WorkloadReportService.php (hunk 4 above)
Comment (jeppekroghitk) [reply src/Service/WorkloadReportService.php:28]: The next PR in line (adds to the workload report functionality), renames the current $viewMode and adds an additional option. To avoid conflicts, i will implements this in the next PR for both the options currently using strings.

https://github.com/itk-dev/economics/pull/129

File: src/Service/WorkloadReportService.php [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+     * @param string $viewMode The view mode (default: 'week')
+     *
+     * @return WorkloadReportData The workload report data
+     *
+     * @throws \Exception When an unexpected value for viewMode is provided
+     */
+    public function getWorkloadReport(string $viewMode = 'week'): WorkloadReportData
+    {
+        // Get period based on viewmode.
+        $periods = match ($viewMode) {
```
Comment (tuj) [inline src/Service/WorkloadReportService.php:31]: I think I would change all match ($viewMode) into one match and do the relevant initialisations for each of the two types.

File: src/DataFixtures/AppFixtures.php [hunk 9]
```diff
@@ -113,9 +124,9 @@ public function load(ObjectManager $manager): void
                         $worklog->setDescription("Beskrivelse af worklog-$key-$i-$j-$k");
                         $worklog->setIsBilled(false);
                         $worklog->setProject($project);
-                        $worklog->setWorker('test@test');
+                        $worklog->setWorker($workerArray[rand(0, 9)]);
```
Comment (tuj) [inline src/DataFixtures/AppFixtures.php:127]: I think it is better to make fixtures stable instead of random.
This makes it easier to write tests.

File: src/DataFixtures/AppFixtures.php [hunk 10]
```diff
@@ -113,9 +124,9 @@ public function load(ObjectManager $manager): void
...
                         $worklog->setTimeSpentSeconds(60 * 15 * ($k + 1));
-                        $worklog->setStarted(new \DateTime());
+                        $worklog->setStarted(\DateTime::createFromFormat('U', (string) rand(strtotime(date('Y-01-01')), strtotime(date('Y-12-31')))));
```
Comment (tuj) [inline src/DataFixtures/AppFixtures.php:129]: Remove randomness

**Review by tuj — CHANGES_REQUESTED** [review]
I think this PR would benefit a lot from adding tests. Especially for testing the logic with calculating workload. Here, it would be great with testing some error cases, edge cases, etc.

**Review by tuj — APPROVED** [review]
